Upgrade react-router

kamilmielnik opened the pull request:
Closes #33832
Closes #16455

### Description

Dependencies:
- upgrade `react-router` to v5
- upgrade `@types/react-router` to v5
- remove explicit dependency on `@types/react-router`
- replace `react-router-redux` with `connected-react-router`
- add `react-router-dom` v5
- add `@types/react-router-dom` v5
- remove `@types/react-router-redux`
- upgrade `history` to v4
- remove `@types/history`
- upgrade `redux-auth-wrapper` to v3
- upgrade `@types/redux-auth-wrapper`

### How to verify

Obviously this is a sensitive change that affects the entire project. Think about edge cases.

App should work exactly the same, except:
- this bug is fixed: #33832
- non-existing partial routes will now correctly render ["Not found" error page](https://github.com/metabase/metabase/assets/6830683/8603eb24-f795-42f0-9670-3aab17791c1c) instead of a blank page (e.g. http://localhost:3000/public/)

kamilmielnik committed:
Upgrade dependencies
- upgrade react-router to v5
- upgrade @types/react-router to v5
- replace react-router-redux with connected-react-router
- add react-router-dom v5
- remove @types/react-router-redux
- upgrade history to v5
- remove @types/history

kamilmielnik committed:
Replace react-router-redux imports with connected-react-router

kamilmielnik committed:
Install @types/react-router-dom

kamilmielnik committed:
Move @types/react-router-dom to devDependencies

kamilmielnik committed:
Use ^ for react-router-dom

kamilmielnik committed:
Remove direct dependency on @types/react-router

kamilmielnik committed:
Update Link imports to use react-router-dom package

kamilmielnik committed:
Update withRouter imports to use react-router-dom package

kamilmielnik committed:
Update Route imports to use react-router-dom package

kamilmielnik committed:
Update Route (now: Location) type imports to use history package

kamilmielnik committed:
Update LocationDescriptor & LocationDescriptorObject imports from history package

kamilmielnik committed:
Update remaining reference to react-router-redux

kamilmielnik committed:
Update router integration with redux store

uladzimirdev commented on frontend/src/metabase/metabot/containers/DatabaseMetabotApp/DatabaseMetabotApp.tsx:
```diff
@@ -37,13 +38,17 @@ const mapStateToProps = (
   { params, location, databases }: RouteProps & DatabaseLoaderProps,
 ): StateProps => {
   const entityId = checkNotNull(extractEntityId(params.databaseId));
+  const query = querystring.parse(location.search);
```
nit: URLSearchParams

kamilmielnik committed:
Downgrade history to v4
- connected-react-router's "isRouter" selector does not allow "action: undefined" in the store

kamilmielnik committed:
Introduce useTrackPageView

kamilmielnik committed:
Use ConnectedRouter in app.js

kamilmielnik committed:
Migrate metabase/routes.jsx part 1

kamilmielnik committed:
Add Switch

kamilmielnik committed:
Migrate splat property

kamilmielnik committed:
Upgrade redux-auth-wrapper to v3

kamilmielnik committed:
Migrate metabase/routes.jsx part 2

kamilmielnik committed:
Upgrade @types/redux-auth-wrapper

kamilmielnik committed:
Refactor metabase/route-guards to TypeScript

kamilmielnik committed:
Migrate auth routing

kamilmielnik committed:
Migrate metabase/routes.jsx part 3

kamilmielnik commented:
Stale for months.
Too much has changed since this PR was in progress.
Closing.
